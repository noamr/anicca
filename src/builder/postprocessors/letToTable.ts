import { Bundle, LetStatement } from '../types'
import {F, S, R, fmap} from './postProcessHelpers'

export default function processLetToTable(bundle: Bundle) : Bundle {
    return fmap(bundle, statement => {
        if (statement.type === 'Let') {
            const tableName = `@let_${statement.name}`
            return [
                S.Table(tableName, { valueType: (statement as LetStatement).valueType}),
                S.Slot(statement.name, { formula: F.get(R(tableName), 0)}),
            ]
        }

        return [statement]
    })
}