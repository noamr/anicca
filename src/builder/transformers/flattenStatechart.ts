import * as _ from 'lodash-es'
import { Bundle, Configuration, ControllerStatement, DispatchAction, FlatStatechart, Formula,
    GotoAction, HistoryConfiguration, Juncture, Modus, State, Statechart, StepResults, Transition, TransitionAction,
 } from '../types'
import {F, R, removeUndefined, S} from './helpers'

const head = <T>(a: T[]) => a.length ? a[0] : null
const flatten = <T>(a: Iterable<Iterable<T>>): Iterable<T> => Array.from(a).reduce((a, b) => [...a, ...b], [])
type Predicate<Ret, Arg> = (a: Arg) => Ret
const some = <T>(a: Iterable<T>, p: Predicate<boolean, T>) => Array.from(a).some(p)
const intersectSets = <T>(a: Set<T>, b: Set<T>) => some(a, x => some(b, y => y === x))
const sort = <T>(a: Iterable<T>, p: (a: T, b: T) => number) => a ? Array.from(a).sort(p) : a

export function flattenState(rootState: State): FlatStatechart {
    const byName: {[name: string]: State} = {}
    const parentMap = new WeakMap<State, State|null>()
    const transitionSourceMap = new WeakMap<Transition, State>()
    const docIndex = new WeakMap<State, number>()
    const allEvents = new Set<string>()

    const getDefault = (state: State): State[] => (state.defaultTargets || []).map(id => byName[id])
    const isParallelState = (s: State) => s.type === 'Parallel'
    const documentIndex = (s: State): number => docIndex.get(s) || 0

    const entryOrder = (a: State, b: State) => documentIndex(a) - documentIndex(b)
    const exitOrder = (a: State, b: State) => documentIndex(b) - documentIndex(a)

    const isAtomicState = (s: State) => s.type === 'State' && !childStates(s).length
    const conditionFor = (t: Transition|null): Formula => (t && t.condition) || trueCondition
    const trueCondition = {$primitive: true} as Formula
    const falseCondition = {$primitive: true} as Formula

    const hashState = (s: State) => `${s.name || `#${documentIndex(s)}`}`
    const hashConfig = (c: Configuration) => sort(c, entryOrder).map(hashState).join(',')
    const hashHistory = (c: HistoryConfiguration) => [...c.keys()].sort(entryOrder).map((hs: State) => `${hashState(hs)}:${hashConfig(c.get(hs) || new Set<State>())}`).join('|')
    const hashModus = (m: Modus): string => `${hashConfig(m.configuration)};${hashHistory(m.history)}`
    const hashJuncture = (j: Juncture|null): string => j ? `${j.event || ''}:${hashModus(j.modus)}` : ''

    const junctures: {[hash: string]: {
        juncture: Juncture|null
        results: StepResults[],
    }} = {}

    const visitedModi = new Set<string>()
    const childStates = (s: State): State[] =>
        ((s.children || []).filter(({type}) => ['State', 'Parallel', 'Final'].includes(type)) as State[])

    const childTransitions = (s: State): Transition[] =>
        ((s.children || []).filter(({type}) => type === 'Transition') as Transition[])

    const getParent = (s: State) => parentMap.get(s) || rootState
    const getTransitionSource = (t: Transition) => transitionSourceMap.get(t) as State

    const nextDocIndex = (next => () => next++)(0)

    resolveState(rootState, null)

    function resolveJuncture(juncture: Juncture|null): void {
        const junctureHash = hashJuncture(juncture)
        if (junctures[junctureHash])
            return

        const results = step(juncture)
        junctures[junctureHash] = {juncture, results}
        for (const r of results) {
            const modusHash = hashModus(r.modus)
            if (visitedModi.has(modusHash))
                continue

            visitedModi.add(modusHash)
            resolveModus(r.modus)
        }
        return
    }

    const resolveModus = (modus: Modus) =>
        [null, ...allEvents].forEach(event => resolveJuncture({event, modus}))

    resolveJuncture(null)
    return {
        events: [...allEvents],
        junctures: new Map(Object.values(junctures).map(({juncture, results}) =>
        ([juncture ? {event: juncture.event, modus: hashModus(juncture.modus)} : null,
            results.map(({condition, execution, modus}) => ({
                condition, execution, modus: hashModus(modus),
            }))] as [Juncture<string>|null, Array<StepResults<string>>]))),
    }

    function resolveTransition(t: Transition, s: State) {
        transitionSourceMap.set(t, s)
        if (t.event)
            allEvents.add(t.event)
    }

    function resolveState(s: State, parent: State|null) {
        byName[s.name] = s
        parentMap.set(s, parent)
        docIndex.set(s, nextDocIndex())

        childStates(s).forEach(child => resolveState(child, s))
        childTransitions(s).forEach(child => resolveTransition(child, s))
    }

    function getAncestors(s1: State): State[] {
        if (!s1 || s1 === rootState)
            return []
        const parent = getParent(s1)
        return [parent, ...getAncestors(parent)].filter(a => a) as State[]
    }

    function getTransitionTargets(t: Transition): State[] {
        const e = (t.actions || []).filter((a: TransitionAction) => a.type === 'Goto')
        return e.map(g => byName[(g as GotoAction).target])
    }

    interface ConditionalTransitionSet {
        condition: Formula
        transitions: Transition[]
    }

    function step(juncture: Juncture|null): StepResults[] {
        if (!juncture) {
            const {modus, execution} = microstep(
                [{type: 'Transition', actions: [{type: 'Goto', target: rootState.name} as GotoAction]}],
                new Set<State>(), new Map())
            return [{
                modus,
                execution,
                condition: trueCondition,
            }]
        }

        const {event, modus: {configuration, history}} = juncture
        const transitionSets = selectTransitions()
        return transitionSets.map(
            t => ({condition: t.condition, ...microstep(t.transitions, configuration, history)}) as StepResults)

        function selectTransitions(): ConditionalTransitionSet[] {
            const atomicStates = [...configuration].filter(isAtomicState).sort(entryOrder)
            const atomicTransitionSets = atomicStates.map(state => {
                const transitions = new Set<Transition>()
                const states = [state, ...getProperAncestors(state, null)]
                for (const s of states)
                    for (const t of childTransitions(s).filter((t => event ? (event === t.event) : !t.event)))
                        transitions.add(t)
                const tlist = [...transitions]
                return tlist.map((t, i) => ({
                    ...t,
                    condition:
                        i === 0 ? conditionFor(t) :
                        i === 1 ? (F.and(F.not(conditionFor(tlist[0])), conditionFor(t))) :
                        F.and(F.not(conditionFor(tlist[0])), conditionFor(t)),
                }) as Transition)
            })

            const computePivot = (current?: Transition[], ...next: Transition[][]): Transition[][] => {
                if (!current)
                    return []
                if (!next.length)
                    return [current]

                return computePivot(...next).map(
                    (p: Transition[]) => [...flatten(current.map(t => ([t, ...p])))] as Transition[])
            }

            const pivot: Transition[][] = computePivot(...atomicTransitionSets)
            return pivot.map((transitions) => ({
                condition:
                    transitions.length === 1 ?
                    conditionFor(head(transitions)) :
                    F.and(...transitions.map(conditionFor)),
                transitions: filterConflicts(transitions),
            }))

            function filterConflicts(selected: Transition[]): Transition[] {
                const filteredTransitions = new Set<Transition>()
                selected.forEach((t1, i) => {
                    let t1Preempted = false
                    const transitionsToRemove = new Set<Transition>()
                    for (const t2 of selected.slice(i + 1)) {
                        if (intersectSets(computeExitSet([t1]), computeExitSet([t2]))) {
                            if (isDescendant(getTransitionSource(t1), getTransitionSource(t2)))
                                transitionsToRemove.add(t2)
                            else
                                t1Preempted = true
                        }
                    }
                    if (!t1Preempted) {
                        for (const t3 of transitionsToRemove)
                            filteredTransitions.delete(t3)
                        filteredTransitions.add(t1)
                    }
                })
                return [...filteredTransitions]
            }
        }

        function getEffectiveTargetStates(t: Transition): State[] {
            let targets: State[] = []
            for (const s of getTransitionTargets(t))
                if (s.type === 'History')
                    if (history.has(s))
                        targets = [...targets, ...(history.get(s) || [])]
                    else
                        targets = [...targets, ...getDefault(s)]
            return [...new Set(targets)]
        }

        function getTransitionDomain(t: Transition) {
            const targets = getEffectiveTargetStates(t)
            const source = getTransitionSource(t)
            if (!targets.length)
                return rootState
            if (isCompound(source) && targets.every(t => isDescendant(t, source)))
                return source
            return findLCCA([source, ...targets]) || rootState
        }

        function computeExitSet(transitions: Transition[]): Set<State> {
            if (!juncture)
                return new Set<State>()

            const statesToExit = new Set<State>()
            transitions
                .filter(t => getTransitionTargets(t).length)
                .forEach(t => {
                    const domain = getTransitionDomain(t)
                    for (const s of configuration)
                        if (!domain || isDescendant(s, domain))
                            statesToExit.add(s)
                })

            return statesToExit
        }

        function microstep( transitions: Transition[],
                            prevConfiguration: Configuration,
                            prevHistory: HistoryConfiguration): {
            modus: Modus,
            execution: TransitionAction[],
        } {
            const configuration = new Set([...prevConfiguration])
            const history = new Map([...prevHistory])
            const onExit = computeStateExit(transitions)
            const onEntry = computeStateEntry(transitions)
            const transitionExec = transitions.map(t => (t.actions || []).filter(a => a.type !== 'Goto'))
                .reduce((a, o) => a.concat(o), [])

            return {
                modus: {configuration, history},
                execution: [...onExit, ...transitionExec, ...onEntry],
            }

            function computeStateExit(transitions: Transition[]): TransitionAction[] {
                let execution: TransitionAction[] = []

                const statesToExit = [...computeExitSet(transitions)].sort(exitOrder)
                for (const s of statesToExit) {
                    for (const h of s.children.filter(c => c.type === 'History') as State[]) {
                        history.set(h, new Set([...configuration].filter(
                            s0 => h.deep
                                ? (isAtomicState(s0) && isDescendant(s0, s))
                                : getParent(s0) === s)))
                    }
                    execution = [...execution, ...(s.onExit || [])]
                    configuration.delete(s)
                }

                return execution
            }

            function computeStateEntry(transitions: Transition[]): TransitionAction[] {
                const statesToEnter = new Set<State>()
                const defaultHistoryContent = new Map<State, TransitionAction[]>()
                const statesForDefaultEntry = new Set<State>()
                let execution: TransitionAction[] = []

                for (const t of transitions) {
                    const targets = getTransitionTargets(t)
                    for (const s of targets)
                        addDescendantStatesToEnter(s)
                    const ancestor = getTransitionDomain(t)
                    for (const s of getEffectiveTargetStates(t))
                        addAncestorStatesToEnter(s, ancestor)
                }

                for (const state of [...statesToEnter].sort(entryOrder)) {
                    configuration.add(state)
                    const onEntry = (state.onEntry || []) as TransitionAction[]
                    const defaultActions = (state.defaultActions || []) as TransitionAction[]
                    execution = [...execution, ...onEntry, ...(statesForDefaultEntry.has(state) ? defaultActions : [])]
                }

                return execution

                function addDescendantStatesToEnter(state: State) {
                    if (state.type === 'History') {
                        const hs = history.get(state)
                        if (hs) {
                            for (const s of hs)
                                addDescendantStatesToEnter(s)
                            for (const s of hs)
                                addAncestorStatesToEnter(s, getParent(s))
                        } else {
                            defaultHistoryContent.set(getParent(state), state.defaultActions || [])
                            for (const s of getDefault(state)) {
                                addDescendantStatesToEnter(s)
                                addAncestorStatesToEnter(s, getParent(state))
                            }
                        }
                        return
                    }
                    statesToEnter.add(state)
                    if (isCompound(state)) {
                        statesForDefaultEntry.add(state)
                        for (const s of getDefault(state)) {
                            addDescendantStatesToEnter(s)
                            addAncestorStatesToEnter(s, state)
                        }
                    } else if (isParallelState(state)) {
                        for (const child of childStates(state))
                            if (!some(statesToEnter, s => isDescendant(s, child)))
                                addDescendantStatesToEnter(child)
                    }
                }

                function addAncestorStatesToEnter(state: State, ancestor: State) {
                    for (const anc of getProperAncestors(state, ancestor)) {
                        statesToEnter.add(anc)
                        if (isParallelState(anc))
                            for (const child of childStates(anc))
                                if (!some(statesToEnter, s => isDescendant(s, child)))
                                    addDescendantStatesToEnter(child)
                    }
                }
            }
        }

        function isCompound(state: State): boolean {
            return state.type === 'State' && !!childStates(state).length
        }

        function isCompoundOrRoot(state: State): boolean {
            return state === rootState || isCompound(state)
        }

        function findLCCA(states: State[]): State {
            if (!states.length)
                return rootState

            for (const anc of getProperAncestors(states[0], null).filter(isCompoundOrRoot)) {
                if (states.slice(1).every(s => isDescendant(s, anc)))
                    return anc
            }

            return rootState
        }

        function isDescendant(s1: State, s2: State) {
            return getAncestors(s1).includes(s2)
        }

        function getProperAncestors(s1: State, s2: State | null) {
            if (s1 === s2)
                return []
            const a = getAncestors(s1)
            if (!s2)
                return a

            const b = getAncestors(s2)
            if (b.indexOf(s1))
                return []

            const s2Index = a.indexOf(s2)
            if (s2Index < 0)
                return a

            return a.slice(0, s2Index)
        }
    }
}
