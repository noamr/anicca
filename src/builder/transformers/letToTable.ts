import { Bundle, LetStatement } from '../types'
import {F, S, R} from './helpers'

export default function transformLetToTable(bundle: Bundle) : Bundle {
    return bundle.flatMap(statement => {
        if (statement.type === 'Let') {
            const tableName = `@let_${statement.name}`
            return [
                S.Table(tableName, { valueType: (statement as LetStatement).valueType}),
                S.Slot(statement.name, { formula: F.get(R(tableName), 0)}),
            ]
        }

        return [statement]
    })
}