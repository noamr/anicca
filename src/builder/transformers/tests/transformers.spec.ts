import letToTable from '../letToTable'
import {flattenState} from '../flattenStatechart'

import {
    parse
} from '../../index'

describe('transformers', () => {
    it('let to table', () => {
        expect(letToTable(parse(`
        let variable: u32
    `))).toEqual(parse(`
        table @let_variable: u32

        slot variable:
            @let_variable[0]
        `, {
            internal: true
        }))
    })

})