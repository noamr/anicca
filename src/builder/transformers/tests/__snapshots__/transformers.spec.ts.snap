// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "outputNames": Object {
    "@view_out": 0,
  },
  "roots": Object {
    "@view_out": 15,
    "idle": 93,
    "outbox": 98,
    "staging": 95,
  },
  "slots": Array [
    Object {
      "value": 0,
    },
    Object {
      "args": Array [
        0,
      ],
      "op": "table",
    },
    Object {
      "args": Array [
        1,
        0,
      ],
      "op": "get",
    },
    Object {
      "args": Array [
        2,
      ],
      "op": "array",
    },
    Object {
      "args": Array [],
      "op": "value",
    },
    Object {
      "value": 4,
    },
    Object {
      "args": Array [
        5,
      ],
      "op": "table",
    },
    Object {
      "args": Array [],
      "op": "key",
    },
    Object {
      "args": Array [
        6,
        7,
      ],
      "op": "get",
    },
    Object {
      "args": Array [
        4,
        8,
      ],
      "op": "eq",
    },
    Object {
      "args": Array [],
      "op": "array",
    },
    Object {
      "args": Array [
        7,
        4,
      ],
      "op": "pair",
    },
    Object {
      "args": Array [
        11,
      ],
      "op": "array",
    },
    Object {
      "args": Array [
        9,
        10,
        12,
      ],
      "op": "cond",
    },
    Object {
      "args": Array [
        3,
        13,
      ],
      "op": "flatMap",
    },
    Object {
      "args": Array [
        14,
      ],
      "op": "encode",
    },
    Object {
      "value": 2,
    },
    Object {
      "args": Array [
        16,
        0,
        16,
      ],
      "op": "array",
    },
    Object {
      "args": Array [
        17,
      ],
      "op": "array",
    },
    Object {
      "args": Array [
        0,
        18,
      ],
      "op": "pair",
    },
    Object {
      "value": 131072,
    },
    Object {
      "args": Array [
        20,
        10,
      ],
      "op": "pair",
    },
    Object {
      "value": 131073,
    },
    Object {
      "value": 1,
    },
    Object {
      "args": Array [
        2,
        23,
      ],
      "op": "plus",
    },
    Object {
      "args": Array [
        0,
        0,
        24,
      ],
      "op": "array",
    },
    Object {
      "args": Array [
        25,
        17,
      ],
      "op": "pair",
    },
    Object {
      "args": Array [
        22,
        26,
      ],
      "op": "pair",
    },
    Object {
      "args": Array [
        19,
        21,
        27,
      ],
      "op": "array",
    },
    Object {
      "args": Array [
        28,
      ],
      "op": "object",
    },
    Object {
      "value": 3,
    },
    Object {
      "args": Array [
        30,
      ],
      "op": "table",
    },
    Object {
      "args": Array [
        31,
      ],
      "op": "size",
    },
    Object {
      "args": Array [
        31,
        0,
      ],
      "op": "get",
    },
    Object {
      "args": Array [
        32,
        33,
        0,
      ],
      "op": "cond",
    },
    Object {
      "args": Array [
        34,
        0,
      ],
      "op": "eq",
    },
    Object {
      "args": Array [
        16,
      ],
      "op": "table",
    },
    Object {
      "args": Array [
        36,
      ],
      "op": "size",
    },
    Object {
      "args": Array [
        36,
        0,
      ],
      "op": "get",
    },
    Object {
      "args": Array [
        37,
        38,
        0,
      ],
      "op": "cond",
    },
    Object {
      "value": 16,
    },
    Object {
      "args": Array [
        39,
        40,
      ],
      "op": "shl",
    },
    Object {
      "args": Array [
        34,
        23,
      ],
      "op": "eq",
    },
    Object {
      "value": null,
    },
    Object {
      "args": Array [
        23,
      ],
      "op": "table",
    },
    Object {
      "args": Array [
        44,
        12,
      ],
      "op": "flatMap",
    },
    Object {
      "args": Array [
        45,
      ],
      "op": "head",
    },
    Object {
      "args": Array [
        44,
        46,
      ],
      "op": "get",
    },
    Object {
      "args": Array [
        42,
        43,
        47,
      ],
      "op": "cond",
    },
    Object {
      "args": Array [
        48,
        43,
      ],
      "op": "neq",
    },
    Object {
      "args": Array [
        48,
      ],
      "op": "head",
    },
    Object {
      "value": 32767,
    },
    Object {
      "args": Array [
        50,
        51,
      ],
      "op": "bwand",
    },
    Object {
      "args": Array [
        49,
        0,
        52,
      ],
      "op": "cond",
    },
    Object {
      "args": Array [
        41,
        53,
      ],
      "op": "bwor",
    },
    Object {
      "args": Array [
        35,
        0,
        54,
      ],
      "op": "cond",
    },
    Object {
      "args": Array [
        29,
        55,
      ],
      "op": "get",
    },
    Object {
      "value": false,
    },
    Object {
      "args": Array [
        57,
      ],
      "op": "not",
    },
    Object {
      "args": Array [
        58,
      ],
      "op": "not",
    },
    Object {
      "args": Array [
        20,
        59,
      ],
      "op": "pair",
    },
    Object {
      "value": true,
    },
    Object {
      "args": Array [
        61,
      ],
      "op": "not",
    },
    Object {
      "args": Array [
        62,
      ],
      "op": "not",
    },
    Object {
      "args": Array [
        22,
        63,
      ],
      "op": "pair",
    },
    Object {
      "args": Array [
        60,
        64,
      ],
      "op": "object",
    },
    Object {
      "args": Array [
        65,
        55,
      ],
      "op": "get",
    },
    Object {
      "args": Array [
        34,
        5,
      ],
      "op": "eq",
    },
    Object {
      "args": Array [
        66,
        66,
        67,
      ],
      "op": "cond",
    },
    Object {
      "args": Array [
        34,
        23,
      ],
      "op": "plus",
    },
    Object {
      "args": Array [
        68,
        23,
        69,
      ],
      "op": "cond",
    },
    Object {
      "args": Array [
        31,
        0,
        70,
      ],
      "op": "put",
    },
    Object {
      "args": Array [
        44,
        46,
      ],
      "op": "delete",
    },
    Object {
      "args": Array [
        71,
        72,
      ],
      "op": "pair",
    },
    Object {
      "args": Array [
        56,
        73,
      ],
      "op": "concat",
    },
    Object {
      "args": Array [
        74,
      ],
      "op": "array",
    },
    Object {
      "args": Array [
        75,
        12,
      ],
      "op": "flatMap",
    },
    Object {
      "args": Array [
        76,
      ],
      "op": "size",
    },
    Object {
      "args": Array [
        76,
      ],
      "op": "head",
    },
    Object {
      "args": Array [
        4,
      ],
      "op": "head",
    },
    Object {
      "value": 1073741823,
    },
    Object {
      "args": Array [
        79,
        80,
      ],
      "op": "bwand",
    },
    Object {
      "value": 15,
    },
    Object {
      "args": Array [
        81,
        82,
      ],
      "op": "shr",
    },
    Object {
      "args": Array [
        83,
        61,
      ],
      "op": "pair",
    },
    Object {
      "args": Array [
        84,
      ],
      "op": "array",
    },
    Object {
      "args": Array [
        7,
        85,
      ],
      "op": "pair",
    },
    Object {
      "args": Array [
        44,
        86,
      ],
      "op": "flatMap",
    },
    Object {
      "args": Array [
        87,
      ],
      "op": "size",
    },
    Object {
      "args": Array [
        87,
      ],
      "op": "head",
    },
    Object {
      "value": -1,
    },
    Object {
      "args": Array [
        88,
        89,
        90,
      ],
      "op": "cond",
    },
    Object {
      "args": Array [
        77,
        78,
        91,
      ],
      "op": "cond",
    },
    Object {
      "args": Array [
        92,
        90,
      ],
      "op": "eq",
    },
    Object {
      "args": Array [
        75,
        92,
      ],
      "op": "get",
    },
    Object {
      "args": Array [
        93,
        10,
        94,
      ],
      "op": "cond",
    },
    Object {
      "args": Array [
        0,
        15,
      ],
      "op": "pair",
    },
    Object {
      "args": Array [
        96,
      ],
      "op": "array",
    },
    Object {
      "args": Array [
        97,
        12,
      ],
      "op": "flatMap",
    },
  ],
  "tableTypes": Object {
    "0": "u32",
    "1": Object {
      "tuple": Array [
        "u32",
        "ByteArray",
      ],
    },
    "2": "u32",
    "3": "u32",
    "4": "string",
  },
}
`;
